File: 2279.maximum-bags-with-full-capacity-of-rocks/solution.kt
import java.util.PriorityQueue
import java.util.Comparator

class Solution {
    fun maximumBags(capacity: IntArray, rocks: IntArray, additionalRocks: Int): Int {
        val compatator = Comparator<Int> {a, b -> a - b}
        val hq = PriorityQueue(capacity.size, compatator)

        capacity.forEachIndexed { i, cap ->
            hq.add(cap - rocks[i])
        }

        var cnt = 0
        var remainRocks = additionalRocks
        while ((remainRocks > 0) && (!hq.isEmpty())) {
            val bag = hq.poll()
            if (remainRocks >= bag) {
                cnt++
            }
            remainRocks -= bag
        }

        return cnt
    }
}

File: 2279.maximum-bags-with-full-capacity-of-rocks/solution.py
class Solution:
    def maximumBags(self, capacity: List[int], rocks: List[int], additionalRocks: int) -> int:
        hq = []
        N = len(capacity)
        for i in range(N):
            heapq.heappush(hq, capacity[i] - rocks[i])

        cnt = 0
        while (additionalRocks > 0) and (len(hq) > 0):
            peek = heapq.heappop(hq)
            if (additionalRocks >= peek):
                cnt += 1
            additionalRocks -= peek

        return cnt
